library/alloc/src/slice.rs

```diff
@@ -1036,149 +1040,150 @@ where
+                merge(v, mid, buf_ptr, is_less);
             }
         }
-
-        // Insert some more elements into the run if it's too short. Insertion sort is faster than
-        // merge sort on short sequences, so this significantly improves performance.
-        while start > 0 && end - start < MIN_RUN {
-            start -= 1;
-            insert_head(&mut v[start..end], &mut is_less);
-        }
-
-        // Push this run onto the stack.
-        runs.push(Run { start, len: end - start });
-        end = start;
-
-        // Merge some pairs of adjacent runs to satisfy the invariants.
-        while let Some(r) = collapse(&runs) {
-            let left = runs[r + 1];
-            let right = runs[r];
-            unsafe {
-                merge(
-                    &mut v[left.start..right.start + right.len],
-                    left.len,
-                    buf.as_mut_ptr(),
-                    &mut is_less,
-                );
-            }
-            runs[r] = Run { start: left.start, len: left.len + right.len };
-            runs.remove(r + 1);
-        }
     }
 
-    // Finally, exactly one run must remain in the stack.
-    debug_assert!(runs.len() == 1 && runs[0].start == 0 && runs[0].len == len);
-
-    // Examines the stack of runs and identifies the next pair of runs to merge. More specifically,
-    // if `Some(r)` is returned, that means `runs[r]` and `runs[r + 1]` must be merged next. If the
-    // algorithm should continue building a new run instead, `None` is returned.
-    //
-    // TimSort is infamous for its buggy implementations, as described here:
-    // http://envisage-project.eu/timsort-specification-and-verification/
-    //
-    // The gist of the story is: we must enforce the invariants on the top four runs on the stack.
-    // Enforcing them on just top three is not sufficient to ensure that the invariants will still
-    // hold for *all* runs in the stack.
-    //
-    // This function correctly checks invariants for the top four runs. Additionally, if the top
-    // run starts at index 0, it will always demand a merge operation until the stack is fully
-    // collapsed, in order to complete the sort.
-    #[inline]
-    fn collapse(runs: &[Run]) -> Option<usize> {
-        let n = runs.len();
-        if n >= 2
-            && (runs[n - 1].start == 0
-                || runs[n - 2].len <= runs[n - 1].len
-                || (n >= 3 && runs[n - 3].len <= runs[n - 2].len + runs[n - 1].len)
-                || (n >= 4 && runs[n - 4].len <= runs[n - 3].len + runs[n - 2].len))
-        {
-            if n >= 3 && runs[n - 3].len < runs[n - 1].len { Some(n - 3) } else { Some(n - 2) }
-        } else {
-            None
+    /// swap contents of left-hand and right-hand slices divided at `mid`
+    ///
+    /// # Safety
+    ///
+    /// `buf_ptr` must point to a slice of `buf` which is long enough to hold `v[mid..]`
+    fn swap_slices<T>(v: &mut [T], mid: usize, buf_ptr: *mut T) {
```

Comment:
Can't you use safe code for this? 
I haven't studied it thoroughly, but would [`swap_with_slice`](https://doc.rust-lang.org/std/primitive.slice.html#method.swap_with_slice) work?

Reply:
`swap_with_slice` only works on slices of equal size.

Reply:
Looking at it, it seems `mid` should really be the middle element. 
So am I right to believe that both subslices would be of equal sizes of off-by-one?
You can then swap the single overflow element manually, no?

Reply:
No, mid can be much greater than the middle element if the slice is pre-sorted past the middle element.

Reply:
> Is this then more a [`rotate`](https://doc.rust-lang.org/nightly/std/primitive.slice.html#method.rotate_left) than a `swap`? A name change might clear things up. I think this is one of the `with_buffer` rotation algorithms, per page 186 of http://elementsofprogramming.com/eop.pdf

Good point, it can be looked at as a large rotate.  Generally, unless the slice is pre-sorted beyond the middle of the slice, `mid` will be in the middle of the slice, so I think of it as a swap of left and right sides of the slice, rather than a large rotate.  Does that make sense?